nomos-da/network/core/src/protocols/replication/behaviour_v2.rs

```diff
@@ -0,0 +1,677 @@
+                        self.outbound_tasks
+                            .push(Box::pin(Self::try_write_message(peer_id, message, stream)));
+
+                        should_wake = true;
+                    }
+                    None => {
+                        self.outbound_streams
+                            .insert(peer_id, StreamState::Idle(stream));
+                    }
+                }
+            }
+            Poll::Ready(Some(Err(error))) => {
+                return Poll::Ready(ToSwarm::GenerateEvent(ReplicationEvent::ReplicationError {
+                    error,
+                }));
+            }
+            _ => {}
+        }
+
+        // Pick the next peer that has a pending message and an idle or unopened stream
+        // and schedule writing a message to it.
+        let next_peer = self.pending_outbound.next_peer(|id| {
+            matches!(
+                self.outbound_streams.get(id),
+                Some(StreamState::Idle(_)) | None
+            )
+        });
+
+        if let Some(peer_id) = next_peer {
+            let message = self
+                .pending_outbound
+                .dequeue_message(&peer_id)
+                .expect("Message is in the queue, ensured by PendingOutbound::next_peer()");
+
+            match self.outbound_streams.entry(peer_id) {
+                Entry::Occupied(mut entry) => match entry.get_mut().take() {
+                    StreamState::Idle(stream) => {
+                        self.outbound_tasks
+                            .push(Box::pin(Self::try_write_message(peer_id, message, stream)));
+                    }
+                    StreamState::Busy => {
+                        unreachable!("The stream is idle, ensured by PendingOutbound::next_peer()")
+                    }
+                },
+                // If there is no stream for this peer yet, try to open one, and then write the
+                // first message into it
+                Entry::Vacant(entry) => {
+                    entry.insert(StreamState::Busy);
+                    let control = self.control.clone();
+                    self.outbound_tasks.push(Box::pin(async move {
+                        let stream = Self::try_open_stream(peer_id, control).await?;
+                        let (peer_id, stream) =
+                            Self::try_write_message(peer_id, message, stream).await?;
+                        Ok((peer_id, stream))
+                    }));
+                }
+            }
+
+            should_wake = true;
+        }
+
+        // Schedule reading from any new incoming streams if possible
+        if let Poll::Ready(Some((peer_id, stream))) = self.incoming_streams.poll_next_unpin(cx) {
+            self.incoming_tasks
+                .push(Self::try_read_message(peer_id, stream).boxed());
+            should_wake = true;
+        }
+
+        if let Some(incoming_message) = incoming_message {
+            return incoming_message;
```

Comment:
The future won't be polled if we have incoming message, I would suggest to change all the `should_wake` with the actual call to `waker().wake_by_ref()` - it can be called multiple times, but the future will be polled only once next time.